pkg/service/storage/ucan.go

```diff
@@ -113,28 +132,60 @@ func NewUCANServer(storageService Service) (server.ServerView, error) {
 						return blob.AcceptOk{}, nil, NewUnsupportedCapabilityError(cap)
 					}
 
-					_, err := storageService.Blobs().Store().Get(ctx, digest)
-					if err != nil {
-						if errors.Is(err, store.ErrNotFound) {
-							return blob.AcceptOk{}, nil, NewAllocatedMemoryNotWrittenError()
+					var loc url.URL
+
+					var forks []fx.Effect
+					if storageService.PDP() == nil {
+						_, err := storageService.Blobs().Store().Get(ctx, digest)
+						if err != nil {
+							if errors.Is(err, store.ErrNotFound) {
+								return blob.AcceptOk{}, nil, NewAllocatedMemoryNotWrittenError()
+							}
+							log.Errorf("getting blob: %w", err)
+							return blob.AcceptOk{}, nil, failure.FromError(err)
 						}
-						log.Errorf("getting blob: %w", err)
-						return blob.AcceptOk{}, nil, failure.FromError(err)
-					}
 
-					loc, err := storageService.Blobs().Access().GetDownloadURL(digest)
-					if err != nil {
-						log.Errorf("creating retrieval URL for blob: %w", err)
-						return blob.AcceptOk{}, nil, failure.FromError(err)
+						loc, err = storageService.Blobs().Access().GetDownloadURL(digest)
+						if err != nil {
+							log.Errorf("creating retrieval URL for blob: %w", err)
+							return blob.AcceptOk{}, nil, failure.FromError(err)
+						}
+					} else {
+						// locate the piece from the pdp service
+						piece, err := storageService.PDP().PieceFinder().FindPiece(ctx, digest, cap.Nb().Blob.Size)
+						if err != nil {
+							log.Errorf("finding piece for blob: %w", err)
+							return blob.AcceptOk{}, nil, failure.FromError(err)
+						}
+						// get a download url
+						loc = storageService.PDP().PieceFinder().URLForPiece(piece)
+						// submit the piece for aggregation
+						err = storageService.PDP().Aggregator().AggregatePiece(ctx, piece)
+						if err != nil {
+							log.Errorf("submitting piece for aggregation: %w", err)
+							return blob.AcceptOk{}, nil, failure.FromError(err)
+						}
+						// generate the invocation that will complete when aggregation is complete and the piece is accepted
+						pieceAccept, err := pdp.PDPAccept.Invoke(
```

Comment:
Musing on this - what if we expose a more generic `blob/retain` that can be used for both? PDP is not really an operation. What I think we really need is an instruction for the blob to continue to be stored over time...and the receipt for an invocation to do this would be details of how to prove it is continually being stored. i.e. a retention receipt would specify the proof type is PDP and contains the inclusion proof. The retention reciept for non-PDP is just "trust me I have it" or "test me with station".

The dictionary definition of "retain" is "continue to have (something); keep possession of".

Side note: perhaps the invocation could include accepted retention policies?

After a blob has been accepted (delivered), the upload service would issue a `blob/retain` invocation to ensure the blob continues to persist on the storage node. Of course for PDP the invocation may have some async tasks to perform (the aggregation), so perhaps simply responds with a `blob/retain/start` effect (not sold on that name).

We can surface this to users by adding the `blob/retain/start` fx to the original `blob/add` invocation.

"retain" also has a nice antonym "release", which we might use to stop the node from submitting PDP proofs (and allow it to garbage collect).

Reply:
to be clear, pdp/accept here is is a LATER thing -- it's saying, this will be done when I've stored successfully in PDP.

Happy to iterate on naming -- I don't love all this as it is.